updated integration_tests.md (#2)
himanshu-kun/machine-controller-manager: docs/development/integration_tests.md

# Integration tests

## Usage

## General setup & configurations

> Integration tests for `machine-controller-manager` and `machine-controller-manager-provider-{provider-name}` can be executed manually by following below steps.

Reviewer: ```suggestion
Integration tests for `machine-controller-manager-provider-{provider-name}` can be executed manually by following below steps.
```

> 1. Clone the repository `machine-controller-manager-provider-{provider-name}` on the local system.

Author: - Generally, the list of steps where the sequence needs to be honoured is always a numbered bullet point. 
- Also, it is advised to avoid as much passive voice statements as possible.
- To use capitalised words for proper nouns like Gardener, Kubernetes and other resource names like MachineClass, Secrets etc.
- As much as possible, one should avoid pronouns like you, we, us, etc and use personas instead. Like, Users, Developers, Customers, etc.

1. Navigate to `machine-controller-manager-provider-{provider-name}` directory and create a `dev` sub-directory in it.

> 1. Copy the kubeconfig of Control Cluster  from into `dev/control-kubeconfig.yaml`. 

Author: The developer who is running these integration tests should necessarily understand what Control and Target Clusters are. So, I would abstract it without details.

1. (optional) Copy the kubeconfig of Target Cluster  into `dev/target-kubeconfig.yaml` and update the `Makefile` variable `TARGET_KUBECONFIG` to point to `dev/target-kubeconfig.yaml`.

> 1. If the tags on instances & associated resources on the provider are of `String` type (for example, GCP tags on its instances are of type `String` and not key-value pair) then add `TAGS_ARE_STRINGS := true` in the `Makefile`.

Reviewer: should we also add that you also need to add a line to export `TAGS_ARE_STRINGS`?

Author: Yes, we should add that. I will add it.

1. Atleast, one of the two controllers' container images must be set in the `Makefile` variables `MCM_IMAGE_TAG` and `MC_IMAGE_TAG` for the controllers to run in the Control Cluster . These images will be used along with `kubernetes/deployment.yaml` to deploy/update controllers in the Control Cluster . If the intention is to run the controllers locally then unset the variables `MCM_IMAGE_TAG` and `MC_IMAGE_TAG` and set variable `MACHINE_CONTROLLER_MANAGER_DEPLOYMENT_NAME := machine-controller-manager` in the `Makefile`.

> 7. If none of the images are specified, `machine-controller-manager` repository will be cloned automatically. Incase, this repository already exists in local system, then create a softlink as below which helps to test changes in `machine-controller-manager` quickly.

Reviewer: `machine-controller-manager` is also cloned for all the cases when control cluster is a shoot cluster irrespective of images being specified.

Author: Yes, I considered this condition and wanted to discuss the approach of MCM IT Suites.

MCM would be cloned when the Control Cluster is a Shoot in order to deploy the CRDs right? Why shouldn't the relevant CRDs be applied from the out of tree repositories only? I mean to say, why should AliCloud MachineClass CRD be applied in an AWS cluster while running MCM AWS Integration tests?

Reviewer: Yes the point is correct, I thought of this too.. But the thing is that there are other steps occurring right after deploying CRDs, like applying machineclass(which is part of MCM IT Suite which itself is called in BeforeSuite()) which would need CRD to be present already. So there is no other place than MCM IT suite as of now to deploy CRD, or we'll have to place `if` conditions if we want to deploy just specific machineclass CRD.

Author: I think you misunderstood my view. You are referring to [this](https://github.com/himanshu-kun/machine-controller-manager/blob/7257227f073ca878aa02a98ce4fd9f2d7ce5f440/pkg/test/integration/common/framework.go#L739-L760) code snippet, I believe. Isn't it? Why should the `kubernetes/crds` from the MCM repo should be used here? Why can't we keep the relevant files in the `MCM-Provider-AWS/kubernetes/crds` folder and apply them in the Control Cluster?

    ```bash
    ln -sf <path-for-machine-controller-manager-repo> dev/mcm
    ```

> ## Scenario based additional configurations

Author: It is better to have sections for easier consumption.

Reviewer: Yep this way its really easy to understand the flow. Thanks

1. (Optional) User can employ a custom `MachineClass` for the tests using below steps:
    1. Deploy a `Secret` named `test-mc-secret` (that contains the provider secret and cloud-config) in the shoot namespace of the Control Cluster. That is, the value of `metadata.namespace` should be `technicalID` of the Shoot and it will be of the pattern `shoot--<project>--<shoot-name>`. Refer [these](https://github.com/gardener/machine-controller-manager/tree/master/kubernetes/machine_classes) `MachineClass` templates for the same. 
    1. Create a `dev/machineclassv1.yaml` file.
        1. `providerSpec.secretRef.name` should refer the secret created in the previous step.
        1. `metadata.namespace` and `providerSpec.secretRef.namespace` should be `technicalID` (`shoot--<project>--<shoot-name>`) of the shoot.
        1.  The name of the `MachineClass` itself should be `test-mc-v1`.

## Running the tests

1. There is a rule `test-integration` in the `Makefile`, which can be used to start the integration test:
    ```bash
    $ make test-integration 
    Starting integration tests...
    Running Suite: Controller Suite
    ===============================
    ```

> 1. The controllers log files (`mcm_process.log` and `mc_process.log`) are stored as temporary files and can be used later.

Reviewer: I have actually changed the location of log files to `.ci/controllers-test/logs` now as I will have to print the logs after the pipeline integration tests run , and its easier to search the logs if it gets stored in the same folder. WDYT?
```suggestion
1. The controllers log files (`mcm_process.log` and `mc_process.log`) are stored in `.ci/controllers-test/logs` repo and can be used later.
```

## Adding Integration Tests for new providers

> For a new provider, Running Integration tests works with no changes. But for the orphan resource test cases to work correctly, the provider-specific API calls and the Resource Tracker Interface (RTI) should be implemented. Please check (`machine-controller-manager-provider-aws`)[https://github.com/gardener/machine-controller-manager-provider-aws/test/integration/provider/rti.go] for reference.

Author: I think it is good to point directly to the file that implements the RTI

Reviewer: I think you have interchanged the brackets while adding the links, just nitpicking.

## Extending integration tests

> - Update [ControllerTests](pkg/test/integration/common/framework.go#L481) to be extend the testcases for all providers. Common testcases for machine|machineDeployment creation|deletion|scaling are packaged into [ControllerTests](pkg/test/integration/common/framework.go#L481).

Author: Update the points to use simple sentences and not passive voice.
